File: go/roman-numerals/roman_numerals.go
package romannumerals

import (
	"errors"
	"strings"
)

type unit struct {
	letter string
	value  int
}

type units []unit

var romanToNumeral = units{
	{"M", 1000},
	{"CM", 900},
	{"D", 500},
	{"CD", 400},
	{"C", 100},
	{"XC", 90},
	{"L", 50},
	{"XL", 40},
	{"X", 10},
	{"IX", 9},
	{"V", 5},
	{"IV", 4},
	{"I", 1},
}

func ToRomanNumeral(number int) (result string, err error) {
	if number <= 0 || number > 3000 {
		return "", errors.New("")
	}

	for _, numeral := range romanToNumeral {
		times := number / numeral.value
		if times > 0 {
			result += strings.Repeat(numeral.letter, times)
			number -= times * numeral.value
		}
	}
	return
}

File: elixir/roman-numerals/lib/roman_numerals.ex
defmodule RomanNumerals do
  @doc """
  Convert the number to a roman number.
  """

  @roman_to_numeral [
    { "M",  1000 },
    { "CM", 900 },
    { "D",  500 },
    { "CD", 400 },
    { "C",  100 },
    { "XC", 90 },
    { "L",  50 },
    { "XL", 40 },
    { "X",  10 },
    { "IX", 9 },
    { "V",  5 },
    { "IV", 4 },
    { "I",  1 }
  ]

  @spec numeral(pos_integer) :: String.t()
  def numeral(number), do: to_roman(number, @roman_to_numeral)

  defp to_roman(0, _), do: ""
  defp to_roman(number, [ { letter, value } | tail ]) do
    times = div(number, value)
    String.duplicate(letter, times) <> to_roman(number - times * value, tail)
  end
end
